src/Nixfmt/Predoc.hs

```diff
@@ -159,9 +158,11 @@ trailing t = [Text 0 0 Trailing t]
 group :: (HasCallStack) => (Pretty a) => a -> Doc
 group x =
   pure . Group RegularG $
-    if p /= [] && (isSoftSpacing (head p) || isSoftSpacing (last p))
-      then error $ "group should not start or end with whitespace, use `group'` if you are sure; " <> show p
-      else p
+    case pretty x of
+      ((hp :<| _) :|> lp)
+        | isSoftSpacing hp || isSoftSpacing lp ->
+            error $ "group should not start or end with whitespace, use `group'` if you are sure; " <> show p
+      _ -> p
```

Comment:
This also changes semantics a bit, since the first pattern only applies with at least 2 elements, whereas before it also applied with just one.

Reply:
Oh you’re right. It was sure a bit late 🙃